server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -128,6 +130,7 @@
   // This variable updated in add(), persist(), and drop()
   private final AtomicInteger rowsCurrentlyInMemory = new AtomicInteger();
   private final AtomicInteger totalRows = new AtomicInteger();
+  private final AtomicLong bytesCurrentlyInMemory = new AtomicLong();
```

Comment:
`rowsCurrentlyInMemory`, `bytesCurrentlyInMemory`, and `totalRows` should be updated in sync because different threads can read/update them and independent update might lead an unexpected behavior. For example, `rowsCurrentlyInMemory` might be already updated, but `bytesCurrentlyInMemory` is not when another thread reads them. 

This should be done when I added `totalRows`, but I missed it. Would you add a class to update all these stats atomically? Like

```java
class MemoryStats
{
  Object lock;
  int rowsInMemory;
  long bytesInMemory;
  int rowsInMemoryAndDisk;

  void add(int rowsInMemory, long bytesInMemory, int rowsInMemoryAndDisk)
  {
    synchronized(lock) {
      this.rowsInMemory += rowsInMemory;
      this.bytesInMemory += bytesInMemory;
      this.rowsInMemoryAndDisk += rowsInMemoryAndDisk;
    }
  }
}
```

Reply:
There's no point updating these in sync if they are not checked in sync, so for this change to be useful, the check should be moved into `MemoryStats` as well. But is it really necessary to check them in sync? The non-`add` threads can only decrease these values. I don't think anything bad will happen if the values decrease while `add` is in the middle of checking them.

What do you think?

Reply:
Ah, yes. The checking part is missing in the above snippet and it should also be in there. 

And yes. It doesn't cause anything bad, but synchronized updates will lead a better behavior like less persisting. Do you have any other concerns?

Reply:
I just want to sort out whether we would be making this change for correctness reasons or aesthetic reasons. If it's for aesthetic reasons, maybe it's not necessary. And if it's for correctness reasons, then we need to make sure we are synchronizing the appropriate amount of things to fix the bug we are trying to fix.

I guess what I'm saying is that it doesn't look to me like "synchronized updates will lead a better behavior like less persisting" is true in practice, since the updates won't be happening simultaneously if the appenderator is being used properly.

It looks to me like this synchronization only useful if `drop` and `add` are called simultaneously [1]. But this shouldn't happen anyway. If it does happen, it will cause worse problems (like data loss, since `drop` is meant to be used after segments have been pushed/handed off. If they haven't been handed off yet then we will lose data).

[1] The counters are updated in threads that call `add`, `persist`, `drop`, and `clear`. But `add`, `persist`, and `clear` already have javadoc saying they must be called from the same thread, so there is no synchronization issue. `drop` is the only one that is allowed to be called from a different thread than the other three.

Reply:
My hypothesis right now is that if `add` and `drop` are called simultaneously for the same segment, a bunch of other bad stuff will happen (not just related to the counters). Luckily, I don't think this happens today, and it would be kind of pointless anyway given that `drop` is used for cleanup after we're done with a segment. IMO, if we are concerned about this case, we should be synchronizing not just the counter updates, but much bigger chunks of the `add` and `drop` methods.

However, I'm still not sure if it's really necessary to be concerned about this case.

Reply:
This is a valid point. I am thinking should I create methods to modify and get these values or you prefer a separate class. Another thing is a read-write lock might be better so multiple threads can read the values.

Reply:
Ok, I got your point. Thanks.

Reply:
@surekhasaharan I think probably it would be ok to leave these alone until we figure out what the right approach is here, it might be a separate PR to either clean up the contract or synchronize some more stuff.

@jihoonson let us know what you think too.

Reply:
@surekhasaharan I talked with @gianm offline. Please see the [new comment](https://github.com/druid-io/druid/pull/5583#discussion_r185358269). If this is the case, `rowsCurrentlyInMemory` and `bytesCurrentlyInMemory ` have no longer to be AtomicInteger because they are used by the same thread.

Reply:
If we end up doing the atomic-to-nonatomic change, I think it makes sense to do it in a separate PR. I think we should at least fix the KafkaIndexTask publishing being in a separate thread first. And get rid of `persist` too.

So my vote is to not worry about it for this particular patch.

Reply:
Agree. @surekhasaharan if you don't want to do in this PR, please raise an issue about it.